### ISSUE TITLE
Can wgmma.async and barrier.arrive Ensure GEMM Completion Before Moving Forward?

### ISSUE THREAD

**Issue by ziyuhuang123** [Dao-AILab/flash-attention#1373]
Since the GEMM here uses wgmma.async, the subsequent barrier.arrive cannot guarantee that GEMM execution is complete, right? It can only ensure that the GEMM has been issued, correct? Otherwise, if strictly following the order of GEMM0, GEMM1, and SOFTMAX, it would be impossible to achieve the overlap within the warp group as shown in the figure below:
![8db1fb4982175e8838169e00b3e44aa](https://github.com/user-attachments/assets/e9ad25e9-385c-4642-a577-d3a5ddfa41a4)

![b90c287986be26ecb9c1296d4f7de16](https://github.com/user-attachments/assets/87e1bd3f-a140-4f02-a0cc-91b7db59255c)

**Comment by ziyuhuang123** [comment]
![40da1808db654e71fba4221de97a851](https://github.com/user-attachments/assets/36dd8bd7-84b2-4bb3-8b70-a7dcc29190eb)
Considering that GEMM is asynchronous (assuming) and softmax is synchronous (execution must complete before proceeding), I tried to combine the intra and inter illustrations to roughly draw something like this...

**Comment by tridao** [comment]
Yes barrier.arrive only guarantees that the gemm has been issued and not necessarily completed
